src/runtime_src/core/pcie/tools/xbmgmt/xspi.cpp

```diff
@@ -2180,3 +2224,127 @@ int XSPI_Flasher::upgradeFirmware2Drv(std::istream& mcsStream0,
+}
+
+int XSPI_Flasher::xclWriteData(std::vector<unsigned char>& data)
+{
+    const int flash_index = 0;
+
+    if (mFlashDev == nullptr)
+        return -EINVAL;
+
+    if (data.size() > xrt_max_data_len) {
+        std::cout << "meta data is too big (" << data.size() << "B) to flash: "
+            << std::endl;
+        return -EINVAL;
+    }
+
+    size_t size = getFlashSize(mDev.get());
+    if (size == 0)
+        return -EINVAL;
+
+    // Write file data onto flash
+    unsigned int addr = size - sizeof(flash_data_header) - data.size();
+    int ret = writeToFlash(mFlashDev, flash_index, addr,
+        data.data(), data.size());
+    if (ret) {
+        std::cout << "failed to write meta data to flash: " << ret << std::endl;
+        return ret;
+    }
+
+    // Write meta data onto flash
+    flash_data_header header;
+    std::memcpy(header.magic, magic.c_str(), magic.size());
+    header.ver = 0;
+    header.data_len = data.size();
+    header.parity = getParity32(data);
+    addr += data.size();
+    ret = writeToFlash(mFlashDev, flash_index, addr,
+        reinterpret_cast<unsigned char *>(&header), sizeof(header));
+    if (ret) {
+        std::cout << "failed to write meata data header to flash: " << ret
+            << std::endl;
+        return ret;
+    }
+
+    std::cout << "Persisted " << data.size() << " bytes of meta data to flash "
+        << flash_index << " @0x" << std::hex << addr << std::dec << std::endl;
+    return 0;
+}
+
+int XSPI_Flasher::xclReadData(std::vector<unsigned char>& data)
+{
+    const int flash_index = 0;
+
+    if (mFlashDev == nullptr)
+        return -EINVAL;
+
+    size_t size = getFlashSize(mDev.get());
+    if (size == 0)
+        return -EINVAL;
+
+    // Read meta data from flash
+    flash_data_header header;
+    unsigned int addr = size - sizeof(header);
+    std::cout << "reading " << sizeof(header) << " bytes of header from flash "
+        << flash_index << " @0x" << std::hex << addr << std::dec << std::endl;
+    int ret = readFromFlash(mFlashDev, flash_index, addr,
+        reinterpret_cast<unsigned char *>(&header), sizeof(header));
+    if (ret) {
+        std::cout << "failed to read header from flash: " << ret << std::endl;
+        return ret;
+    }
+    if (std::memcmp(header.magic, magic.c_str(), magic.size())) {
+        std::cout << "bad header magic: " << header.magic << std::endl;
+        return -EINVAL;
+    }
+
+    // Read file data from flash
+    data.resize(header.data_len);
+    addr -= header.data_len;
+    std::cout << "reading " << header.data_len << " bytes of data from flash "
+        << flash_index << " @0x" << std::hex << addr << std::dec << std::endl;
+    ret = readFromFlash(mFlashDev, flash_index, addr, data.data(), data.size());
+    if (ret) {
+        std::cout << "failed to read data from flash: " << ret << std::endl;
+        return ret;
+    }
+    if (header.parity ^ getParity32(data)) {
+        std::cout << "data is corrupted" << std::endl;
```

Comment:
More description should be added here.  For example: "ERROR:  XRT Flash Data is corrupted. Expected: 0xFFEF, Actual: 0xABDF

Reply:
will do.